openscm/adapters/__init__.py

```diff
@@ -115,64 +164,162 @@ def step(self) -> np.datetime64:
         self._step()
         return self._current_time
 
+    def _set_model_from_parameters(self):
+        update_time_points = self._timeseries_time_points_require_update()
+        for name, view in self._get_view_iterator():
+            pv = self._parameter_views[name]
+            update_time = pv.parameter_type in (
+                ParameterType.POINT_TIMESERIES,
+                ParameterType.AVERAGE_TIMESERIES,
+            ) and (update_time_points or pv.empty)
+            if update_time:
+                current_view = self._parameter_views[name]
+                unit = current_view.unit
+                region = current_view.region
+                timeseries_type = current_view.parameter_type
+
+                self._parameter_views[name] = self._parameters.timeseries(
+                    name,
+                    unit,
+                    time_points=self._get_time_points(timeseries_type),
+                    region=region,
+                    timeseries_type=timeseries_type,
+                    interpolation="linear",  # TODO: take these from ParameterSet
+                    extrapolation="none",
+                )
+
+            update_para = not self._parameter_views[name].empty and (
+                self._parameter_versions[name] < view.version or update_time
+            )
+            if update_para:
+                if isinstance(name, tuple) and name[0] == self.name:
+                    self._direct_set = True
+                self._update_model_parameter(name)
+                self._parameter_versions[name] = self._parameter_views[name].version
+                self._direct_set = False
+
+    def _get_view_iterator(self):
+        view_iterator = self._parameter_views.items()
+        view_iterator = sorted(view_iterator, key=lambda s: len(s[0]))
+        generic_views = [
+            v
+            for v in view_iterator
+            if not isinstance(v[1], dict)
+            and v[1].parameter_type == ParameterType.GENERIC
+        ]
+        scalar_views = [
+            v
+            for v in view_iterator
+            if not isinstance(v[1], dict)
+            and v[1].parameter_type == ParameterType.SCALAR
+        ]
+        other_views = [
+            v
+            for v in view_iterator
+            if isinstance(v[1], dict)
+            or v[1].parameter_type not in (ParameterType.GENERIC, ParameterType.SCALAR)
+        ]
+        return generic_views + scalar_views + other_views
+
+    def _update_model_parameter(self, name: HierarchicalName) -> None:
+        para = self._parameter_views[name]
+        self._update_model(name, para)
+
+    @staticmethod  # should we unify getters?
```

Comment:
Why have a separate function for that anyway? (same for setters)

Reply:
Because timeseries views require you to look at `.values` whereas scalar views require you to look at `.value` and having the check of which attribute you need to look at is a pain so I just put it in one place. If we unified the getters (i.e. made them both `.value` or `.values` we could avoid this).